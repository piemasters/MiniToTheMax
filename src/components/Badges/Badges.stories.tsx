import type { Meta, StoryObj } from '@storybook/react-vite';

import { Badges } from './Badges';

const meta: Meta<typeof Badges> = {
  component: Badges,
};

export default meta;
type Story = StoryObj<typeof Badges>;

export const Primary: Story = {
  render: () => <Badges />,
};
