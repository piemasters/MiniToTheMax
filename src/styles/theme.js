export default {
  colors: {
    text: '#000',
    muted: '#999',
  },
};
