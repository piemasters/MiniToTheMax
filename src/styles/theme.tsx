export interface Theme {
  colors: {
    text: string;
    muted: string;
  };
}

const appTheme: Theme = {
  colors: {
    text: '#000',
    muted: '#999',
  },
};

export default appTheme;
