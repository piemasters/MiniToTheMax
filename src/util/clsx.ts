/* eslint-disable @typescript-eslint/no-unused-expressions */
/* eslint-disable @typescript-eslint/no-explicit-any */
export type ClassValue =
  | ClassArray
  | ClassDictionary
  | string
  | number
  | null
  | boolean
  | undefined;
type ClassDictionary = Record<string, any>;
type ClassArray = ClassValue[];

function toVal(mix: string | number | boolean | unknown[] | ClassDictionary) {
  let k,
    y,
    str = '';

  if (typeof mix === 'string' || typeof mix === 'number') {
    str += mix;
  } else if (typeof mix === 'object') {
    if (Array.isArray(mix)) {
      const len = mix.length;
      for (k = 0; k < len; k++) {
        if (mix[k]) {
          if ((y = toVal(mix[k]))) {
            str && (str += ' ');
            str += y;
          }
        }
      }
    } else {
      for (y in mix) {
        if (mix[y]) {
          str && (str += ' ');
          str += y;
        }
      }
    }
  }

  return str;
}

export function clsx(...args: ClassValue[]): string {
  let i = 0,
    tmp,
    x,
    str = '';
  const len = args.length;
  for (; i < len; i++) {
    if ((tmp = args[i])) {
      if ((x = toVal(tmp))) {
        str && (str += ' ');
        str += x;
      }
    }
  }
  return str;
}

export default clsx;
