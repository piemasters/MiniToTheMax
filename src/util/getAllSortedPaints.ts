import { PaintDetails } from '../types';
import { textToId } from './textToId';
import { sortPaints } from './sortPaints';
import {
  airPaints,
  basePaints,
  contrastPaints,
  dryPaints,
  layerPaints,
  shadePaints,
  sprayPaints,
  technicalPaints,
} from '../data/paints/citadel';
import { gameColorPaints } from '../data/paints';

export const getAllSortedPaints = (): PaintDetails[] => {
  const allPaints: PaintDetails[] = [
    // citadel
    ...airPaints,
    ...basePaints,
    ...contrastPaints,
    ...dryPaints,
    ...layerPaints,
    ...shadePaints,
    ...sprayPaints,
    ...technicalPaints,
    // vallejo
    ...gameColorPaints,
  ];

  // Remove duplicates using a Set
  const uniquePaints = Array.from(
    new Set(
      allPaints.map(
        (paint) =>
          `${textToId(paint.company)}_${textToId(paint.name)}_${textToId(paint.category)}`
      )
    )
  ).map((uniqueKey) =>
    allPaints.find(
      (paint) =>
        `${textToId(paint.company)}_${textToId(paint.name)}_${textToId(paint.category)}` ===
        uniqueKey
    )
  ) as PaintDetails[];

  // Sort the paints using the sortPaints function
  return uniquePaints.sort(sortPaints);
};
