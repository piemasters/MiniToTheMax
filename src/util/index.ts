export * from './pagination';
export * from './paintFilters';
