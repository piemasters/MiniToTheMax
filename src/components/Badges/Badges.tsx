import type { FC } from 'react';
import { css, useTheme } from '@emotion/react';
import kebabCase from 'lodash.kebabcase';

import { Badge } from '../Badge';
import { PageLink } from '../PageLink';
import { BadgeNames } from '../../types';
import { Theme } from '../../styles/theme';

/**
 * The Badges component is collection of badges with the most popular categories to display on the homepage.
 */
export const Badges: FC = () => {
  const theme = useTheme() as Theme;

  const badges: BadgeNames[] = [
    BadgeNames.showcase,
    BadgeNames.tutorials,
    BadgeNames.reviews,
    BadgeNames.battleReports,
    //BadgeNames.build,
  ];

  const wrapperStyles = css`
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
    justify-content: center;
  `;

  const linkStyles = css`
    color: ${theme.colors.textSecondary};
    display: flex;
    text-decoration: none;
    flex-direction: column;
    opacity: 0.9;
    &:hover {
      opacity: 1;
    }
  `;

  return (
    <div css={wrapperStyles} data-testid="badges">
      {badges.map((badge: BadgeNames) => {
        return (
          <PageLink
            to={kebabCase(badge)}
            type={'cover'}
            direction={'up'}
            linkStyle={linkStyles}
            key={badge}
          >
            <Badge type={badge} />
          </PageLink>
        );
      })}
    </div>
  );
};

export default Badges;
