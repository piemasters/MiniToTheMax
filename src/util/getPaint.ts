import {
  CitadelPaintCategories,
  PaintCategory,
  PaintDetails,
  VallejoPaintCategories,
} from '../types';
import { getAllSortedPaints } from './getAllSortedPaints';

const allPaints = getAllSortedPaints();

const paintCategoryPriority = [
  ...CitadelPaintCategories,
  ...VallejoPaintCategories,
] as const;

type category = (typeof paintCategoryPriority)[number];

export const getPaint = (
  name: string,
  category?: PaintCategory
): { paint?: PaintDetails; error?: string; type?: string } => {
  const paintMatches = allPaints.filter((paint) => paint.name === name);

  if (paintMatches.length === 0) {
    return { error: 'Incorrect Paint Name', type: name };
  }

  if (paintMatches.length === 1) {
    return { paint: paintMatches[0] };
  }

  if (category) {
    const categoryMatches = paintMatches.filter(
      (paint) => paint.category === category
    );
    if (categoryMatches.length === 0) {
      return { error: 'Incorrect Paint Category', type: category };
    }
    return { paint: categoryMatches[0] };
  }

  const getCategoryPriority = (category: PaintCategory) => {
    const index = paintCategoryPriority.indexOf(category);
    return index === -1 ? Infinity : index;
  };
  paintMatches.sort(
    (a, b) => getCategoryPriority(a.category) - getCategoryPriority(b.category)
  );

  return { paint: paintMatches[0] };
};
