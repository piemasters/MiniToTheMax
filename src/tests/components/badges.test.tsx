import React from 'react';
import Badges from '../../components/badges';
import { getByTestId, render } from '@testing-library/react';
import { ThemeProvider } from 'emotion-theming';
import theme from '../../styles/theme';
import { renderWithTransitionProvider } from '../util/transition-provider';

describe('Badges', () => {
  test('renders correctly', () => {
    const { container } = renderWithTransitionProvider(
      <ThemeProvider theme={theme}>
        <Badges />
      </ThemeProvider>
    );

    expect(getByTestId(container, 'badges')).toBeTruthy();
    expect(container).toMatchSnapshot();
  });
});
