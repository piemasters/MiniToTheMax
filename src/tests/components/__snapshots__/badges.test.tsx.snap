// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badges renders correctly 1`] = `
<div>
  <div
    css="You have tried to stringify object returned from \`css\` function. It isn't supposed to be used directly (e.g. as value of the \`className\` prop), but rather handed to emotion so it can handle it (e.g. as value of \`css\` prop)."
    data-testid="badges"
  >
    <div
      css="You have tried to stringify object returned from \`css\` function. It isn't supposed to be used directly (e.g. as value of the \`className\` prop), but rather handed to emotion so it can handle it (e.g. as value of \`css\` prop)."
      data-testid="badge"
    >
      <a
        bg="#eb1d23"
        css="You have tried to stringify object returned from \`css\` function. It isn't supposed to be used directly (e.g. as value of the \`className\` prop), but rather handed to emotion so it can handle it (e.g. as value of \`css\` prop)."
        data-testid="page-link-cover"
        direction="up"
        duration="1"
        href="/showcase"
      >
        <img
          alt=""
          css="You have tried to stringify object returned from \`css\` function. It isn't supposed to be used directly (e.g. as value of the \`className\` prop), but rather handed to emotion so it can handle it (e.g. as value of \`css\` prop)."
          src="test-file-stub"
        />
        <p
          css="You have tried to stringify object returned from \`css\` function. It isn't supposed to be used directly (e.g. as value of the \`className\` prop), but rather handed to emotion so it can handle it (e.g. as value of \`css\` prop)."
        >
          showcase
        </p>
      </a>
    </div>
    <div
      css="You have tried to stringify object returned from \`css\` function. It isn't supposed to be used directly (e.g. as value of the \`className\` prop), but rather handed to emotion so it can handle it (e.g. as value of \`css\` prop)."
      data-testid="badge"
    >
      <a
        bg="#eb1d23"
        css="You have tried to stringify object returned from \`css\` function. It isn't supposed to be used directly (e.g. as value of the \`className\` prop), but rather handed to emotion so it can handle it (e.g. as value of \`css\` prop)."
        data-testid="page-link-cover"
        direction="up"
        duration="1"
        href="/tutorials"
      >
        <img
          alt=""
          css="You have tried to stringify object returned from \`css\` function. It isn't supposed to be used directly (e.g. as value of the \`className\` prop), but rather handed to emotion so it can handle it (e.g. as value of \`css\` prop)."
          src="test-file-stub"
        />
        <p
          css="You have tried to stringify object returned from \`css\` function. It isn't supposed to be used directly (e.g. as value of the \`className\` prop), but rather handed to emotion so it can handle it (e.g. as value of \`css\` prop)."
        >
          tutorials
        </p>
      </a>
    </div>
    <div
      css="You have tried to stringify object returned from \`css\` function. It isn't supposed to be used directly (e.g. as value of the \`className\` prop), but rather handed to emotion so it can handle it (e.g. as value of \`css\` prop)."
      data-testid="badge"
    >
      <a
        bg="#eb1d23"
        css="You have tried to stringify object returned from \`css\` function. It isn't supposed to be used directly (e.g. as value of the \`className\` prop), but rather handed to emotion so it can handle it (e.g. as value of \`css\` prop)."
        data-testid="page-link-cover"
        direction="up"
        duration="1"
        href="/reviews"
      >
        <img
          alt=""
          css="You have tried to stringify object returned from \`css\` function. It isn't supposed to be used directly (e.g. as value of the \`className\` prop), but rather handed to emotion so it can handle it (e.g. as value of \`css\` prop)."
          src="test-file-stub"
        />
        <p
          css="You have tried to stringify object returned from \`css\` function. It isn't supposed to be used directly (e.g. as value of the \`className\` prop), but rather handed to emotion so it can handle it (e.g. as value of \`css\` prop)."
        >
          reviews
        </p>
      </a>
    </div>
    <div
      css="You have tried to stringify object returned from \`css\` function. It isn't supposed to be used directly (e.g. as value of the \`className\` prop), but rather handed to emotion so it can handle it (e.g. as value of \`css\` prop)."
      data-testid="badge"
    >
      <a
        bg="#eb1d23"
        css="You have tried to stringify object returned from \`css\` function. It isn't supposed to be used directly (e.g. as value of the \`className\` prop), but rather handed to emotion so it can handle it (e.g. as value of \`css\` prop)."
        data-testid="page-link-cover"
        direction="up"
        duration="1"
        href="/battle-reports"
      >
        <img
          alt=""
          css="You have tried to stringify object returned from \`css\` function. It isn't supposed to be used directly (e.g. as value of the \`className\` prop), but rather handed to emotion so it can handle it (e.g. as value of \`css\` prop)."
          src="test-file-stub"
        />
        <p
          css="You have tried to stringify object returned from \`css\` function. It isn't supposed to be used directly (e.g. as value of the \`className\` prop), but rather handed to emotion so it can handle it (e.g. as value of \`css\` prop)."
        >
          battle Reports
        </p>
      </a>
    </div>
  </div>
</div>
`;
